agent/xds/delta.go

```diff
@@ -0,0 +1,728 @@
+			newResourceMap, err := indexResources(newRes)
+			if err != nil {
+				return err
+			}
+			newVersions, err := computeResourceVersions(newResourceMap)
+			if err != nil {
+				return err
+			}
+
+			resourceMap = newResourceMap
+			currentVersions = newVersions
+			ready = true
+
+		case <-retryTimer:
+		}
+
+		// It doesn't matter why, we can reset this timer since we're doing the
+		// state machine again now.
+		retryTimer = nil
+
+		// Trigger state machine
+		switch state {
+		case stateDeltaInit:
+			if node == nil {
+				// This can't happen (tm) since stateCh is nil until after the first req
+				// is received but lets not panic about it.
+				continue
+			}
+			// Start authentication process, we need the proxyID
+			proxyID = structs.NewServiceID(node.Id, parseEnterpriseMeta(node))
+
+			// Start watching config for that proxy
+			stateCh, watchCancel = s.CfgMgr.Watch(proxyID)
+			// Note that in this case we _intend_ the defer to only be triggered when
+			// this whole process method ends (i.e. when streaming RPC aborts) not at
+			// the end of the current loop iteration. We have to do it in the loop
+			// here since we can't start watching until we get to this state in the
+			// state machine.
+			defer watchCancel()
+
+			g.Logger = g.Logger.With("service_id", proxyID.String()) // enhance future logs
+
+			g.Logger.Trace("watching proxy, pending initial proxycfg snapshot for xDS")
+
+			// Now wait for the config so we can check ACL
+			state = stateDeltaPendingInitialConfig
+		case stateDeltaPendingInitialConfig:
+			if cfgSnap == nil {
+				// Nothing we can do until we get the initial config
+				continue
+			}
+
+			// Got config, try to authenticate next.
+			state = stateDeltaRunning
+
+			// Upgrade the logger
+			switch cfgSnap.Kind {
+			case structs.ServiceKindConnectProxy:
+			case structs.ServiceKindTerminatingGateway:
+				g.Logger = g.Logger.Named(logging.TerminatingGateway)
+			case structs.ServiceKindMeshGateway:
+				g.Logger = g.Logger.Named(logging.MeshGateway)
+			case structs.ServiceKindIngressGateway:
+				g.Logger = g.Logger.Named(logging.IngressGateway)
+			}
+
+			g.Logger.Trace("Got initial config snapshot")
+
+			// Lets actually process the config we just got or we'll mis responding
+			fallthrough
+		case stateDeltaRunning:
+			// Check ACLs on every Discovery{Request,Response}.
+			if err := checkStreamACLs(cfgSnap); err != nil {
+				return err
+			}
+			// For the first time through the state machine, this is when the
+			// timer is first started.
+			extendAuthTimer()
+
+			if !ready {
+				g.Logger.Trace("Skipping delta computation because we haven't gotten a snapshot yet")
+				continue
+			}
+
+			var pendingTypes []string
+			for typeUrl, handler := range handlers {
+				if !handler.registered {
+					continue
+				}
+				if len(handler.pendingUpdates) > 0 {
+					pendingTypes = append(pendingTypes, typeUrl)
+				}
+			}
+			if len(pendingTypes) > 0 {
```

Comment:
When we compute what is "wrong" in envoy's representation we build a list of clusters, endpoints, routes, and listeners to add/remove/update and send up to 4 responses at once in a particular order so as to not drop traffic.

Until ALL of the outstanding updates are {N,}ACKed by envoy we do not send additional changes yet so we don't get ahead of ourselves and break something. There are likely some subset of these changes that can go out simultaneously but that's tricky to figure out from these opaque blobs so this approach is the simplest for now.